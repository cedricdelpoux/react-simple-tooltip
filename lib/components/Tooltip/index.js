"use strict"

exports.__esModule = true

var _templateObject = _taggedTemplateLiteralLoose(
    ["\n  0% {\n      opacity: 0;\n  }\n  100% {\n      opacity: 1;\n  }\n"],
    ["\n  0% {\n      opacity: 0;\n  }\n  100% {\n      opacity: 1;\n  }\n"]
  ),
  _templateObject2 = _taggedTemplateLiteralLoose(
    ["\n  animation: ", "ms ", " 0s 1 ", ";\n"],
    ["\n  animation: ", "ms ", " 0s 1 ", ";\n"]
  ),
  _templateObject3 = _taggedTemplateLiteralLoose(
    ["\n  position: absolute;\n  ", ";\n  ", ";\n"],
    ["\n  position: absolute;\n  ", ";\n  ", ";\n"]
  ),
  _templateObject4 = _taggedTemplateLiteralLoose(
    [
      "\n  ",
      ";\n  bottom: 100%;\n  left: 50%;\n  transform: translateX(-50%);\n  margin-bottom: ",
      "px;\n",
    ],
    [
      "\n  ",
      ";\n  bottom: 100%;\n  left: 50%;\n  transform: translateX(-50%);\n  margin-bottom: ",
      "px;\n",
    ]
  ),
  _templateObject5 = _taggedTemplateLiteralLoose(
    [
      "\n  ",
      ";\n  top: 100%;\n  left: 50%;\n  transform: translateX(-50%);\n  margin-top: ",
      "px;\n",
    ],
    [
      "\n  ",
      ";\n  top: 100%;\n  left: 50%;\n  transform: translateX(-50%);\n  margin-top: ",
      "px;\n",
    ]
  ),
  _templateObject6 = _taggedTemplateLiteralLoose(
    [
      "\n  ",
      ";\n  right: 100%;\n  top: 50%;\n  transform: translateY(-50%);\n  margin-right: ",
      "px;\n",
    ],
    [
      "\n  ",
      ";\n  right: 100%;\n  top: 50%;\n  transform: translateY(-50%);\n  margin-right: ",
      "px;\n",
    ]
  ),
  _templateObject7 = _taggedTemplateLiteralLoose(
    [
      "\n  ",
      ";\n  left: 100%;\n  top: 50%;\n  transform: translateY(-50%);\n  margin-left: ",
      "px;\n",
    ],
    [
      "\n  ",
      ";\n  left: 100%;\n  top: 50%;\n  transform: translateY(-50%);\n  margin-left: ",
      "px;\n",
    ]
  )

var _react = require("react")

var _react2 = _interopRequireDefault(_react)

var _propTypes = require("prop-types")

var _propTypes2 = _interopRequireDefault(_propTypes)

var _core = require("@emotion/core")

var _propTypes3 = require("../../utils/propTypes")

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {default: obj}
}

function _objectWithoutProperties(obj, keys) {
  var target = {}
  for (var i in obj) {
    if (keys.indexOf(i) >= 0) continue
    if (!Object.prototype.hasOwnProperty.call(obj, i)) continue
    target[i] = obj[i]
  }
  return target
}

function _taggedTemplateLiteralLoose(strings, raw) {
  strings.raw = raw
  return strings
} /** @jsx jsx */
/* eslint-disable react/prop-types */
/* eslint-disable no-unused-vars */
/* eslint-disable react/react-in-jsx-scope */
/* eslint-disable react/display-name */

var fadeAnimation = (0, _core.keyframes)(_templateObject)

var animation = function animation(props) {
  return (0, _core.css)(
    _templateObject2,
    props.fadeDuration,
    props.fadeEasing,
    fadeAnimation
  )
}

// prettier-ignore
// eslint-disable-next-line no-unused-vars
var Base = function Base(props) {
  return (0, _core.css)(_templateObject3, props.fadeDuration && props.fadeDuration > 0 && animation(props), props.zIndex && "z-index: " + props.zIndex + ";");
};

var Top = function Top(props) {
  return (0, _core.css)(_templateObject4, Base(props), props.offset)
}

var Bottom = function Bottom(props) {
  return (0, _core.css)(_templateObject5, Base(props), props.offset)
}

var Left = function Left(props) {
  return (0, _core.css)(_templateObject6, Base(props), props.offset)
}

var Right = function Right(props) {
  return (0, _core.css)(_templateObject7, Base(props), props.offset)
}

var tooltips = {
  left: function left(_ref) {
    var children = _ref.children,
      props = _objectWithoutProperties(_ref, ["children"])

    return (0, _core.jsx)("div", {css: Left(props)}, children)
  },
  top: function top(_ref2) {
    var children = _ref2.children,
      props = _objectWithoutProperties(_ref2, ["children"])

    return (0, _core.jsx)("div", {css: Top(props)}, children)
  },
  right: function right(_ref3) {
    var children = _ref3.children,
      props = _objectWithoutProperties(_ref3, ["children"])

    return (0, _core.jsx)("div", {css: Right(props)}, children)
  },
  bottom: function bottom(_ref4) {
    var children = _ref4.children,
      props = _objectWithoutProperties(_ref4, ["children"])

    return (0, _core.jsx)("div", {css: Bottom(props)}, children)
  },
}

var Tooltip = function Tooltip(_ref5) {
  var children = _ref5.children,
    offset = _ref5.offset,
    open = _ref5.open,
    placement = _ref5.placement,
    zIndex = _ref5.zIndex,
    fadeDuration = _ref5.fadeDuration,
    fadeEasing = _ref5.fadeEasing

  var Component = tooltips[placement] || tooltips.top
  return (
    open &&
    (0, _core.jsx)(
      Component,
      {
        offset: offset,
        zIndex: zIndex,
        fadeDuration: fadeDuration,
        fadeEasing: fadeEasing,
      },
      children
    )
  )
}

Tooltip.propTypes =
  process.env.NODE_ENV !== "production"
    ? {
        children: _propTypes2.default.any.isRequired,
        offset: _propTypes2.default.number,
        open: _propTypes2.default.bool,
        placement: _propTypes2.default.string,
        zIndex: _propTypes2.default.number,
        fadeEasing: _propTypes3.easingPropType,
        fadeDuration: _propTypes2.default.number,
      }
    : {}

exports.default = Tooltip
module.exports = exports["default"]
