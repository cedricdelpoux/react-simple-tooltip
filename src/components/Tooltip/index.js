/** @jsx jsx */
import PropTypes from "prop-types"
import {css, keyframes, jsx} from "@emotion/core"
import {easingPropType} from "../../utils/propTypes"

const fadeAnimation = keyframes`
  0% {
      opacity: 0;
  }
  100% {
      opacity: 1;
  }
`

const animation = props => css`
  animation: ${props.fadeDuration}ms ${props.fadeEasing} 0s 1 ${fadeAnimation};
`

// prettier-ignore
const Base = (props) => css`
  position: absolute;
  ${props.fadeDuration && props.fadeDuration > 0 && animation(props)};
  ${props.zIndex && `z-index: ${props.zIndex};`};
`

const Top = props => css`
  ${Base(props)};
  bottom: 100%;
  left: 50%;
  transform: translateX(-50%);
  margin-bottom: ${props.offset}px;
`

const Bottom = props => css`
  ${Base(props)};
  top: 100%;
  left: 50%;
  transform: translateX(-50%);
  margin-top: ${props.offset}px;
`

const Left = props => css`
  ${Base(props)};
  right: 100%;
  top: 50%;
  transform: translateY(-50%);
  margin-right: ${props.offset}px;
`

const Right = props => css`
  ${Base(props)};
  left: 100%;
  top: 50%;
  transform: translateY(-50%);
  margin-left: ${props.offset}px;
`

const BaseToolTop = ({fn, children, ...props}) => (
  <div css={fn(props)}>{children}</div>
)

BaseToolTop.propTypes = {
  fn: PropTypes.func.isRequired,
  children: PropTypes.any.isRequired,
  offset: PropTypes.number,
  open: PropTypes.bool,
  zIndex: PropTypes.number,
  fadeEasing: easingPropType,
  fadeDuration: PropTypes.number,
}

const tooltips = {
  left: ({children, ...props}) => BaseToolTop({fn: Left, children, ...props}),
  top: ({children, ...props}) => BaseToolTop({fn: Top, children, ...props}),
  right: ({children, ...props}) => BaseToolTop({fn: Right, children, ...props}),
  bottom: ({children, ...props}) =>
    BaseToolTop({fn: Bottom, children, ...props}),
}

const Tooltip = ({
  children,
  offset,
  open,
  placement,
  zIndex,
  fadeDuration,
  fadeEasing,
}) => {
  const Component = tooltips[placement] || tooltips.top
  return (
    open && (
      <Component
        offset={offset}
        zIndex={zIndex}
        fadeDuration={fadeDuration}
        fadeEasing={fadeEasing}
      >
        {children}
      </Component>
    )
  )
}

Tooltip.propTypes = {
  children: PropTypes.any.isRequired,
  offset: PropTypes.number,
  open: PropTypes.bool,
  placement: PropTypes.string,
  zIndex: PropTypes.number,
  fadeEasing: easingPropType,
  fadeDuration: PropTypes.number,
}

export default Tooltip
