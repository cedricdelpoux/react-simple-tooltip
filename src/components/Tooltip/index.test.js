import {mount} from "enzyme"
import React from "react"
import Tooltip from "./index"
import * as emotion from "emotion"
import {matchers, createSerializer} from "jest-emotion"

expect.extend(matchers)
expect.addSnapshotSerializer(createSerializer(emotion))

const tooltipProps = {
  offset: 8,
  zIndex: 8,
  open: true,
  children: "ðŸ˜Ž",
}
const TooltipUpFixture = <Tooltip placement="bottom" {...tooltipProps} />
const TooltipBottomFixture = <Tooltip placement="top" {...tooltipProps} />
const TooltipLeftFixture = <Tooltip placement="right" {...tooltipProps} />
const TooltipRightFixture = <Tooltip placement="left" {...tooltipProps} />
const NoTooltipFixture = <Tooltip {...tooltipProps} open={false} />

describe("Tooltip", () => {
  it("renders", () => {
    const wrappers = [
      mount(TooltipUpFixture),
      mount(TooltipBottomFixture),
      mount(TooltipLeftFixture),
      mount(TooltipRightFixture),
    ]

    wrappers.forEach(wrapper => {
      expect(wrapper).not.toHaveStyleRule("animation")
      expect(wrapper).toMatchSnapshot()
    })
  })

  it("do not render", () => {
    expect(
      mount(NoTooltipFixture)
        .children()
        .get(0)
    ).toBeFalsy()
  })

  it("should create a Tooltip with an animation", () => {
    const wrapper = mount(
      <Tooltip {...tooltipProps} fadeDuration={100} fadeEasing={"ease-out"} />
    )
    expect(wrapper).toHaveStyleRule("animation", /100ms ease-out 0s 1 \w+/)
    expect(wrapper).toMatchSnapshot()
  })
})
