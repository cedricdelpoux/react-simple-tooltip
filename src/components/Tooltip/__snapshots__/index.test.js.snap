// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders 1`] = `
.emotion-0 {
  position: absolute;
  z-index: 8;
  top: 100%;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  margin-top: 8px;
}

.emotion-0 {
  position: absolute;
  z-index: 8;
  top: 100%;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  margin-top: 8px;
}

<Tooltip
  offset={8}
  open={true}
  placement="bottom"
  zIndex={8}
>
  <bottom
    offset={8}
    zIndex={8}
  >
    <EmotionCssPropInternal
      __EMOTION_LABEL_PLEASE_DO_NOT_USE__="BaseToolTop"
      __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
      css={
        Object {
          "map": undefined,
          "name": "10m8uvq",
          "next": undefined,
          "styles": "
          
          position: absolute;
          ;
          z-index: 8;;
        ;
          top: 100%;
          left: 50%;
          transform: translateX(-50%);
          margin-top: 8px;
        ",
        }
      }
    >
      <div
        className="emotion-0"
      >
        ðŸ˜Ž
      </div>
    </EmotionCssPropInternal>
  </bottom>
</Tooltip>
`;

exports[`Tooltip renders 2`] = `
.emotion-0 {
  position: absolute;
  z-index: 8;
  bottom: 100%;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  margin-bottom: 8px;
}

.emotion-0 {
  position: absolute;
  z-index: 8;
  bottom: 100%;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  margin-bottom: 8px;
}

<Tooltip
  offset={8}
  open={true}
  placement="top"
  zIndex={8}
>
  <top
    offset={8}
    zIndex={8}
  >
    <EmotionCssPropInternal
      __EMOTION_LABEL_PLEASE_DO_NOT_USE__="BaseToolTop"
      __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
      css={
        Object {
          "map": undefined,
          "name": "158wz1b",
          "next": undefined,
          "styles": "
          
          position: absolute;
          ;
          z-index: 8;;
        ;
          bottom: 100%;
          left: 50%;
          transform: translateX(-50%);
          margin-bottom: 8px;
        ",
        }
      }
    >
      <div
        className="emotion-0"
      >
        ðŸ˜Ž
      </div>
    </EmotionCssPropInternal>
  </top>
</Tooltip>
`;

exports[`Tooltip renders 3`] = `
.emotion-0 {
  position: absolute;
  z-index: 8;
  left: 100%;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  margin-left: 8px;
}

.emotion-0 {
  position: absolute;
  z-index: 8;
  left: 100%;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  margin-left: 8px;
}

<Tooltip
  offset={8}
  open={true}
  placement="right"
  zIndex={8}
>
  <right
    offset={8}
    zIndex={8}
  >
    <EmotionCssPropInternal
      __EMOTION_LABEL_PLEASE_DO_NOT_USE__="BaseToolTop"
      __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
      css={
        Object {
          "map": undefined,
          "name": "1il8gpc",
          "next": undefined,
          "styles": "
          
          position: absolute;
          ;
          z-index: 8;;
        ;
          left: 100%;
          top: 50%;
          transform: translateY(-50%);
          margin-left: 8px;
        ",
        }
      }
    >
      <div
        className="emotion-0"
      >
        ðŸ˜Ž
      </div>
    </EmotionCssPropInternal>
  </right>
</Tooltip>
`;

exports[`Tooltip renders 4`] = `
.emotion-0 {
  position: absolute;
  z-index: 8;
  right: 100%;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  margin-right: 8px;
}

.emotion-0 {
  position: absolute;
  z-index: 8;
  right: 100%;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  margin-right: 8px;
}

<Tooltip
  offset={8}
  open={true}
  placement="left"
  zIndex={8}
>
  <left
    offset={8}
    zIndex={8}
  >
    <EmotionCssPropInternal
      __EMOTION_LABEL_PLEASE_DO_NOT_USE__="BaseToolTop"
      __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
      css={
        Object {
          "map": undefined,
          "name": "1ey0qyy",
          "next": undefined,
          "styles": "
          
          position: absolute;
          ;
          z-index: 8;;
        ;
          right: 100%;
          top: 50%;
          transform: translateY(-50%);
          margin-right: 8px;
        ",
        }
      }
    >
      <div
        className="emotion-0"
      >
        ðŸ˜Ž
      </div>
    </EmotionCssPropInternal>
  </left>
</Tooltip>
`;

exports[`Tooltip should create a Tooltip with an animation 1`] = `
@keyframes animation-0 {
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
}

.emotion-0 {
  position: absolute;
  -webkit-animation: 100ms ease-out 0s 1 animation-0;
  animation: 100ms ease-out 0s 1 animation-0;
  z-index: 8;
  bottom: 100%;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  margin-bottom: 8px;
}

.emotion-0 {
  position: absolute;
  -webkit-animation: 100ms ease-out 0s 1 animation-0;
  animation: 100ms ease-out 0s 1 animation-0;
  z-index: 8;
  bottom: 100%;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  margin-bottom: 8px;
}

<Tooltip
  fadeDuration={100}
  fadeEasing="ease-out"
  offset={8}
  open={true}
  zIndex={8}
>
  <top
    fadeDuration={100}
    fadeEasing="ease-out"
    offset={8}
    zIndex={8}
  >
    <EmotionCssPropInternal
      __EMOTION_LABEL_PLEASE_DO_NOT_USE__="BaseToolTop"
      __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
      css={
        Object {
          "map": undefined,
          "name": "13wl4ay",
          "next": Object {
            "name": "animation-1f15dgw",
            "next": undefined,
            "styles": "@keyframes animation-1f15dgw{
          0% {
              opacity: 0;
          }
          100% {
              opacity: 1;
          }
        }",
          },
          "styles": "
          
          position: absolute;
          
          animation: 100ms ease-out 0s 1 animation-1f15dgw;
        ;
          z-index: 8;;
        ;
          bottom: 100%;
          left: 50%;
          transform: translateX(-50%);
          margin-bottom: 8px;
        ",
        }
      }
    >
      <div
        className="emotion-0"
      >
        ðŸ˜Ž
      </div>
    </EmotionCssPropInternal>
  </top>
</Tooltip>
`;
